catalog/cataloger_merge.go

```diff
@@ -128,128 +165,99 @@ func hasCommitDifferences(tx db.Tx, leftID, rightID int64) (bool, error) {
+	// collect changes to apply  on the branch
+	paths := make([]string, 0, MergeBatchSize)
+	ctidArray := make([]string, 0, MergeBatchSize)
+	var tombstonePaths []string
+	for _, diffRec := range mergeBatch {
+		if diffRec.DiffType == DifferenceTypeRemoved || diffRec.DiffType == DifferenceTypeChanged {
+			paths = append(paths, diffRec.Entry.Path)
+		}
+		if (diffRec.DiffType == DifferenceTypeAdded || diffRec.DiffType == DifferenceTypeChanged) &&
+			diffRec.EntryCtid != nil {
+			ctidArray = append(ctidArray, *diffRec.EntryCtid)
+		}
+		if diffRec.DiffType == DifferenceTypeRemoved && diffRec.parentTombstoneNeeded {
+			tombstonePaths = append(tombstonePaths, diffRec.Entry.Path)
+		}
 	}
-
-	// do the merge based on the relation
-	previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
-	if err != nil {
-		return 0, err
+	// apply changes
+	if len(paths) > 0 {
+		// set entries that exist in the right branch as deleted by entries that were removed or changed
+		setMaxCommit := sq.Update("catalog_entries").
+			Set("max_commit", previousMaxCommitID).
+			Where("branch_id = ? and max_commit = ?", rightID, MaxCommitID).
+			Where(sq.Eq{"path": paths})
+		sql, args, err := setMaxCommit.PlaceholderFormat(sq.Dollar).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sql, args...)
+		if err != nil {
+			return err
+		}
 	}
-	switch relation {
-	case RelationTypeFromParent:
-		err = c.mergeFromParent(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeFromChild:
-		err = c.mergeFromChild(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	case RelationTypeNotDirect:
-		err = c.mergeNonDirect(ctx, tx, previousMaxCommitID, nextCommitID, leftID, rightID, committer, msg, metadata)
-	default:
-		return 0, ErrUnsupportedRelation
+	if len(ctidArray) > 0 {
+		// copy entries from left to right
+		internalSelect := sq.Select(int64Str(rightID), "path", "physical_address", "creation_date", "size", "checksum", "metadata", int64Str(int64(nextCommitID))).
+			From("catalog_entries").
+			Where(sq.Eq{"ctid": ctidArray})
+		copyEntries := sq.Insert("catalog_entries").
+			Columns("branch_id", "path", "physical_address", "creation_date", "size", "checksum", "metadata", "min_commit").
+			Select(internalSelect)
+		sql, args, err := copyEntries.PlaceholderFormat(sq.Dollar).ToSql()
+		if err != nil {
+			return err
+		}
+		_, err = tx.Exec(sql, args...)
+		if err != nil {
+			return err
+		}
 	}
-	if err != nil {
-		return 0, err
+	// insert tombsotnes into parent branch that has a removed entry in its lineage
+	if len(tombstonePaths) > 0 {
+		values := "(VALUES ('" + strings.Join(tombstonePaths, "'),('") + "')) AS t(path)"
+		sql := `INSERT INTO catalog_entries (branch_id,path,physical_address,size,checksum,metadata,min_commit,max_commit)
+				SELECT $1,path,'',0,'','{}',$2,0 FROM ` + values
+		_, err := tx.Exec(sql, rightID, previousMaxCommitID)
```

Comment:
why not use squirrel in this case - preferred over the string concat

Reply:
It is hard to tell without further details.  However if I might guess, interpolating a `StringArray` gives you VALUES of an row shape and you would like it to have a columns shape.  One way around it is:
```sql
ariels@postgres=# select unnest(array [2,3,4,5]);
 unnest 
--------
      2
      3
      4
      5
(4 rows)
```

Another is to use `CopyFrom` support (which is kinda why we have to use `pgx` anyway).